### PR TITLE
feat(autonomy): CEO-facing approval management skills (#428)

### LINKED ISSUE #428: feat: approval management skills (approve, deny, dismiss, list)
## Background

ADR-018 defines three CEO-facing skills for resolving pending approval
requests, plus a read-only skill for listing them. This issue implements
the "CEO responds" side of the approval flow.

**Depends on:** #427 (gate-level approval trigger — `pending_approval`
rows must exist in `autonomy_action_log` before these skills can operate on them).

### 1. `approve-action` skill

- `action_risk: "none"`, `sensitivity: "elevated"` (CEO-only)
- Input: `short_ref` (optional when only one pending request exists)
- Looks up the `pending_approval` row by `short_ref`. When only one pending
  row exists and no `short_ref` is provided, resolves to the sole row. When
  multiple exist and reference is ambiguous, returns an error listing all
  pending requests with their `short_ref` and `description`.
- Transitions row to `outcome = 'approved'`, sets `resolved_at`,
  `resolved_by = 'ceo'`
- Re-executes the original skill with the stored `payload`, passing
  `humanApproved: true` to `ExecutionLayer.invoke()` (and threaded through
  to `OutboundGateway` where applicable)
- Writes a new `autonomy_action_log` row for the re-execution with `parent_action_id`
  pointing to the approved row
- Publishes `human.decision` audit event
- Returns the re-execution result to the coordinator (so the CEO gets
  immediate feedback if the re-execution fails due to stale context)

### 2. `deny-action` skill

- `action_risk: "none"`, `sensitivity: "elevated"` (CEO-only)
- Input: `short_ref` (same resolution logic as `approve-action`)
- Transitions row to `outcome = 'denied'`, sets `resolved_at`,
  `resolved_by = 'ceo'`
- Publishes `human.decision` audit event
- No re-execution

### 3. `dismiss-action` skill

- `action_risk: "none"`, `sensitivity: "elevated"` (CEO-only)
- Input: `short_ref` (same resolution logic as `approve-action`)
- Transitions row to `outcome = 'resolved_externally'`, sets `resolved_at`,
  `resolved_by = 'ceo'`
- For use when the CEO handled the action outside Curia
- Publishes `human.decision` audit event

### 4. `list-pending-actions` skill

- `action_risk: "none"`, `sensitivity: "elevated"` (CEO-only)
- Returns all `pending_approval` rows that have not expired, with
  `short_ref`, `description`, `skill_name`, `created_at`, `expires_at`
- Used by the coordinator to present pending requests to the CEO and
  to resolve ambiguous references

### 5. `send-draft` integration

When `send-draft` successfully sends a draft that has a corresponding
`pending_approval` row in `autonomy_action_log` (linked by draft ID in `payload`),
transition that row to `outcome = 'approved'` and publish `human.decision`.
This captures the approval decision for Phase 3 scoring without requiring
the CEO to separately call `approve-action`.

### 6. Pin skills to coordinator

Add `approve-action`, `deny-action`, `dismiss-action`, and
`list-pending-actions` to `pinned_skills` in `agents/coordinator.yaml`.

## Acceptance Criteria

- [ ] `approve-action`: resolves `short_ref`, transitions row, re-executes with `humanApproved`, writes child `autonomy_action_log` row, publishes `human.decision`
- [ ] `approve-action`: returns re-execution result to coordinator (success or failure)
- [ ] `approve-action`: ambiguous `short_ref` returns error listing pending requests
- [ ] `approve-action`: sole pending row resolves without explicit `short_ref`
- [ ] `deny-action`: transitions row to `denied`, publishes `human.decision`
- [ ] `dismiss-action`: transitions row to `resolved_externally`, publishes `human.decision`
- [ ] `list-pending-actions`: returns all non-expired `pending_approval` rows
- [ ] `send-draft`: transitions corresponding `autonomy_action_log` row on successful send
- [ ] All four skills are `sensitivity: "elevated"` (CEO-only)
- [ ] All four skills pinned to coordinator
- [ ] Handler tests for each skill covering happy path, edge cases, and security gate

### PR DESCRIPTION
**Pull request by josephfung**
## Summary

- Adds four CEO-facing skills for resolving pending approval requests: `approve-action`, `deny-action`, `dismiss-action`, `list-pending-actions`
- Fixes the `short_ref` race condition from the #436 review: adds `UNIQUE(task_id, short_ref)` constraint (migration 032) and a retry loop in `ApprovalTriggerService.request()` that catches Postgres error 23505
- Adds `resolvePending()`, `findAllPending()`, and `resolveRow()` to `ActionLogRepo`, plus a `ResolveResult` discriminated union for safe branch exhaustion
- Wires two new capabilities (`executionLayer`, `actionLogRepo`) through the skill platform; adds `'dismiss'` to `HumanDecisionPayload.decision`
- Pins all four skills in the coordinator and adds a Pending Approval Requests guidance section to the coordinator prompt

## Test Plan

- [ ] All 1879 unit tests pass (`npm test`)
- [ ] Typecheck clean (`npm run typecheck`)
- [ ] Trigger an autonomy gate block (high-risk skill) → `list-pending-actions` returns the pending row with correct `short_ref` and description
- [ ] `approve-action <short_ref>` → blocked skill re-executes with `humanApproved: true`, child row linked via `parent_action_id`, `human.decision` audit event published
- [ ] `deny-action <short_ref>` → row transitions to `denied`, `human.decision` event published, no re-execution
- [ ] `dismiss-action <short_ref>` → row transitions to `resolved_externally`, `human.decision` event with `decision: 'dismiss'`
- [ ] Omitting `short_ref` when exactly one request is pending → resolved automatically
- [ ] Omitting `short_ref` when multiple are pending → ambiguous error with list
- [ ] Non-CEO caller → rejected by all four skills
- [ ] Run migration 032 against test DB and verify constraint blocks duplicate `(task_id, short_ref)` inserts

## Notes

- `scoring-pass.ts` has a pre-existing issue where `callLlmJudge` doesn't validate the parsed JSON fields before use (can silently store all-zero scores). Out of scope for this PR — tracked separately.
- `senderId`/`channelId` fallback to `'unknown'` in audit events is a low-risk edge case (only occurs under misconfiguration, since all four skills require `ceoInitiated: true`).

Closes #428

## Release Notes

* **New Features**
  * Added CEO approval workflow: approve, deny, dismiss, or list pending action requests.
  * Approved actions now trigger automatic re-execution of previously blocked skills.
  * New audit event tracking for human decision dismissals.

* **Documentation**
  * Added approval flow design specification and implementation guide.

### CODE REVIEW

**Comment by josephfung** [comment]
@coderabbitai review

**Comment by josephfung** [comment]
@coderabbitai review

File: src/skills/execution.ts [hunk 1]
```diff
@@ -482,6 +485,10 @@ export class ExecutionLayer {
       browserService: this.browserService,
       bullpenService: this.bullpenService,
       executiveProfileService: this.executiveProfileService,
+      actionLogRepo: this.actionLogRepo,
+      // executionLayer injects `this` so approve-action can re-invoke blocked skills
+      // with humanApproved: true (see ADR-018). Only approve-action should declare this.
+      executionLayer: this,
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/skills/execution.ts:491]: Fixed in efff49c. Added a hard runtime guard in `execution.ts` that checks `manifest.name !== 'approve-action'` before injecting `executionLayer`. Any other skill declaring this capability gets a fail-closed refusal with a SECURITY-level log entry. This is the enforcement boundary — the manifest comment was advisory only.

File: skills/approve-action/handler.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      row.skillName,
+      row.payload,
+      ctx.caller,
+      {
+        humanApproved: true,
+        taskEventId: ctx.taskEventId,
+        conversationId: row.conversationId ?? undefined,
+      },
+    );
+
```
[the remark this answers is not in the document]
Comment (josephfung) [reply skills/approve-action/handler.ts]: Fixed in efff49c. `resolveRow()` now returns `Promise<boolean>`. `approve-action` checks the return value and aborts re-execution when `false` (concurrent resolution). Same guard added to `deny-action` and `dismiss-action`. New test: `aborts re-execution when resolveRow returns false` verifies `executionLayer.invoke` is never called.

File: skills/approve-action/handler.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      success: true,
+      data: {
+        approved: row.shortRef,
+        description: row.description,
+        reExecutionSuccess: reResult.success,
+        reExecutionResult: reResult.success ? reResult.data : (reResult as { error: string }).error,
+      },
+    };
+  }
+}
```
[the remark this answers is not in the document]
Comment (josephfung) [reply skills/approve-action/handler.ts:137]: Fixed in efff49c. Added top-level try/catch wrapping the main handler body (after CEO/capability validation). All unguarded awaits are now caught and converted to `{ success: false, error }` per the skill contract.

File: src/autonomy/action-log-repo.ts [hunk 4]
```diff
@@ -176,6 +178,113 @@ export class ActionLogRepo {
...
+       WHERE id = $1 AND outcome = 'pending_approval'`,
+      [id, outcome, resolvedBy],
+    );
+    if ((result.rowCount ?? 0) === 0) {
+      // Row was already resolved by a concurrent call — idempotent no-op.
+      this.logger.warn({ id, outcome, resolvedBy }, 'action-log-repo: resolveRow affected 0 rows — row may have been resolved concurrently');
+    } else {
+      this.logger.debug({ id, outcome, resolvedBy }, 'action-log-repo: row resolved');
+    }
+  }
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/autonomy/action-log-repo.ts:224]: Fixed in efff49c. `resolveRow()` now returns `Promise<boolean>` — `true` when the row was actually transitioned, `false` when another actor resolved first. All three handler callers (approve, deny, dismiss) check the return and abort with a descriptive error on `false`. New test in `action-log-repo.test.ts` verifies the boolean return for both cases.

File: skills/deny-action/handler.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+          decision: 'deny',
+          deciderId: senderId,
+          deciderChannel: channelId,
+          subjectEventId: row.taskId,
+          subjectSummary: `CEO denied: ${row.description ?? row.skillName}`,
+          contextShown: ['short_ref', 'description', 'skill_name'],
+          presentedAt: row.createdAt,
+          decidedAt: new Date(),
+          defaultAction: 'block',
+          parentEventId: ctx.taskEventId ?? '',
```
[the remark this answers is not in the document]
Comment (josephfung) [reply skills/deny-action/handler.ts]: Fixed in efff49c. All three handlers now guard the publish block with `if (ctx.taskEventId)` — when absent, we log a warning and skip the audit event rather than emitting a broken empty parentEventId. This matches the pattern: if we cannot form valid lineage, do not emit a misleading event.

File: skills/list-pending-actions/handler.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      ctx.log.warn('list-pending-actions: rejected — ceoInitiated flag absent or false');
+      return { success: false, error: 'This skill requires direct CEO authorization.' };
+    }
+
+    if (!ctx.actionLogRepo) {
+      return { success: false, error: 'list-pending-actions requires actionLogRepo capability' };
+    }
+
+    const rows = await ctx.actionLogRepo.findAllPending();
+
```
[the remark this answers is not in the document]
Comment (josephfung) [reply skills/list-pending-actions/handler.ts]: Fixed in efff49c. `findAllPending()` call is now wrapped in try/catch — database errors return `{ success: false, error }` per the skill contract. Test added: `returns SkillResult error when findAllPending throws`.

File: skills/list-pending-actions/handler.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    const pending = rows.map((row) => ({
+      short_ref: row.shortRef,
+      description: row.description,
+      skill_name: row.skillName,
+      created_at: row.createdAt.toISOString(),
+      expires_at: row.expiresAt?.toISOString() ?? null,
+    }));
```
[the remark this answers is not in the document]
Comment (josephfung) [reply skills/list-pending-actions/handler.ts]: Fixed in efff49c. Now uses `toLocalIso(Math.floor(date.getTime() / 1000), ctx.timezone)` for both `created_at` and `expires_at`, matching the established calendar skill pattern. Added `displayTimezone: formatDisplayTimezone(ctx.timezone, new Date())` to result data. Tests updated to verify presence of `displayTimezone`.

File: src/autonomy/action-log-repo.ts [hunk 8]
```diff
@@ -176,6 +178,113 @@ export class ActionLogRepo {
...
+         WHERE short_ref = $1
+           AND outcome = 'pending_approval'
+           AND expires_at > now()
+         ORDER BY created_at ASC
+         LIMIT 1`,
+        [shortRef],
+      );
+      if (pending.rows.length > 0) {
+        return { found: true, row: mapRow(pending.rows[0]) };
+      }
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/autonomy/action-log-repo.ts:253]: Fixed in efff49c. Changed Query 1 from `LIMIT 1` to `LIMIT 2`. When 2+ pending rows match the same `short_ref` across different tasks, we now return `ambiguous` with a descriptive error listing both matches, instead of silently picking the oldest.

File: src/db/migrations/032_unique_task_short_ref.sql [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+--
+-- Adds a UNIQUE constraint on (task_id, short_ref) to prevent race conditions
+-- in short_ref generation. Null short_ref values (non-approval rows) do not
+-- violate the constraint — Postgres treats nulls as never equal.
+--
+-- See issue #428 and the #436 review discussion.
+
+ALTER TABLE autonomy_action_log
+  ADD CONSTRAINT uq_aal_task_short_ref UNIQUE (task_id, short_ref);
+
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/db/migrations/032_unique_task_short_ref.sql:32]: Fixed in efff49c. Added a `DO $$ ... END $$` pre-check that queries for duplicate `(task_id, short_ref)` pairs and raises a clear exception with remediation SQL if any exist. The table was introduced 1-2 days ago so duplicates are unlikely, but this prevents a confusing bare ALTER TABLE failure.

File: skills/approve-action/handler.test.ts [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+function makeCtx(overrides?: Partial<SkillContext>): SkillContext {
+  return {
+    input: { short_ref: 'cal-1' },
+    secret: () => '',
+    log: createSilentLogger(),
+    taskMetadata: { ceoInitiated: true, senderId: 'ceo-1', channelId: 'cli' },
+    taskEventId: 'task-1',
+    caller: { contactId: 'ceo-1', role: 'ceo', channel: 'cli' },
+    ...overrides,
+  } as SkillContext;
```
[the remark this answers is not in the document]
Comment (josephfung) [reply skills/approve-action/handler.test.ts:43]: Fixed in efff49c. All four test files now use `secret: (name: string) => { throw new Error(\`secret '${name}' not configured in test\`); }` to match production behavior where `ctx.secret()` throws when a secret is not set.

File: skills/approve-action/handler.test.ts [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    invoke: vi.fn().mockResolvedValue(result ?? { success: true, data: { event_id: 'evt-123' } }),
+  } as unknown as ExecutionLayer;
+}
+
+describe('ApproveActionHandler', () => {
+  it('rejects non-CEO callers', async () => {
+    const handler = new ApproveActionHandler();
+    const result = await handler.execute(makeCtx({ taskMetadata: {} }));
+    expect(result.success).toBe(false);
+  });
```
[the remark this answers is not in the document]
Comment (josephfung) [reply skills/approve-action/handler.test.ts:75]: Fixed in efff49c. The non-CEO test now overrides both `taskMetadata: {}` and `caller: { contactId: 'user-99', role: 'contact', channel: 'email' }` — exercises both the `ceoInitiated` flag check and would catch a future regression if the guard shifts to caller identity.
